### PR TITLE
Ensure a snapshot exists before returning it

### PR DESCRIPTION
**Pull request by brianmichel**
The Firebase SDK always creates a `DocumentSnapshot` and returns it when you call into the get APIs, we need to ensure we call `exists` which was pointed out when we filed this issue https://github.com/firebase/firebase-cpp-sdk/issues/1493#issuecomment-1828444320.

### CODE REVIEW

**Comment by brianmichel** [comment]
Current dependencies on/for this PR:
* main
  * **PR #11** <a href="https://app.graphite.dev/github/pr/thebrowsercompany/swift-firebase/11?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/thebrowsercompany/swift-firebase/11?utm_source=stack-comment).

File: Sources/FirebaseFirestore/DocumentReference+Swift.swift [hunk 1]
```diff
@@ -48,7 +48,7 @@ extension DocumentReference {
       future.Wait(firebase.FutureBase.kWaitTimeoutInfinite)
     }
 
-    return snapshot.pointee.is_valid() ? snapshot.pointee : nil
+    return snapshot.pointee.exists && snapshot.pointee.is_valid() ? snapshot.pointee : nil
```
Comment (compnerd) [inline Sources/FirebaseFirestore/DocumentReference+Swift.swift]: How about guarding that the snapshot exists first?

```suggestion
    guard snapshot.pointee.exists else { return nil }
    return snapshot.pointee.is_valid() ? snapshot.pointee : nil
```

File: Sources/firebase/include/FirebaseFirestore.hh [hunk 2]
```diff
@@ -64,6 +64,11 @@ snapshot_reference(const ::firebase::firestore::DocumentSnapshot snapshot) {
   return snapshot.reference();
 }
 
+inline bool
+snapshot_exists(const ::firebase::firestore::DocumentSnapshot snapshot) {
```
Comment (compnerd) [inline Sources/firebase/include/FirebaseFirestore.hh]: Please take this by const reference to avoid a copy.

```suggestion
snapshot_exists(const ::firebase::firestore::DocumentSnapshot &snapshot) {
```

File: Sources/firebase/include/FirebaseFirestore.hh [hunk 3]
```diff
@@ -101,7 +106,7 @@ field_value_workaround(::firebase::firestore::MapFieldValue value) {
 
 inline ::firebase::firestore::DocumentReference
 collection_document(::firebase::firestore::CollectionReference collection,
-                    std::string &document_path) {
+                    std::string document_path) {
```
Comment (compnerd) [inline Sources/firebase/include/FirebaseFirestore.hh]: Why the change to make this a non-const value?

Comment (brianmichel) [reply Sources/firebase/include/FirebaseFirestore.hh]: I pushed a commit to put this back, but github hasn't picked it up yet :(

**Review by compnerd — APPROVED** [review]
